13647 fixed wrong feature flag for checking admin fine-grained authz

sschu opened the pull request:
Fixes https://github.com/keycloak/keycloak/issues/13647.

sschu committed:
13647 fixed wrong feature flag for checking admin fine-grained authz

sschu commented:
@pedroigor 

Here is the PR for fixing other wrong usages of the feature flag.

pedroigor commented:
@sschu Thanks. Looks like some tests are failing because now the `authz` field in `MgmtPermissions` is null and no check exists if this reference is null.

sschu commented:
I'll have a look.

sschu commented:
@pedroigor I had to adapted some tests to also enable the profile feature correctly. I hope that does the trick.

pedroigor commented:
@sschu Thanks again! Makes more sense for me too.

However, I'm afraid we still have some null checks to make like:

* https://github.com/keycloak/keycloak/blob/d3b43a9f59764b763bffc11ba1f087f929f433ef/services/src/main/java/org/keycloak/services/resources/admin/permissions/RolePermissions.java#L126.
* https://github.com/keycloak/keycloak/blob/d3b43a9f59764b763bffc11ba1f087f929f433ef/services/src/main/java/org/keycloak/services/resources/admin/permissions/RolePermissions.java#L524

The only reason why they don't fail is because of checks like https://github.com/keycloak/keycloak/blob/c3ea847b3eb2d9081279ed168c2a5704216f3139/services/src/main/java/org/keycloak/services/resources/admin/permissions/RolePermissionManagement.java#L36.

So the places still using the `authz` even though `null` are only called if permissions are enabled for a specific area.

sschu committed:
13647 Added null checks and some comments/questions for discussions. Will be squashed later if accepted.

sschu commented:
@pedroigor 
I also saw the the code for the `XyzPermissions` classes is not very safe and relies on being used in a safe way from external. I did not like that so I added another commit for easier review adding checks. I am somewhat confident I caught all cases. 

The general idea is to only check for the profile feature in one place in 'MgmtPermissions' and use `authz!=null` as an equivalent internal check down the road (not sure on this one, maybe checking for the feature directly is more explicit - but its also more code to read). In general `ADMIN_FINE_GRAINED_AUTHZ` enabled should be equivalent to `authz!=null`. Also, 'resourceServer!=null' means 'authz!=null' so it is safe to make calls to it. I am not 100% sure the reverse is also true so maybe there are cases where `resourceServer==null` and `authz!=null`. But most checks just check  `resourceServer!=null` and can safely use `authz` afterwards.

I also added some source code comments/questions, feel free to respond to them. If you like the changes I would finalise the second commit and finally squash it before merging.

pedroigor commented on services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java:
```diff
@@ -292,6 +293,7 @@ public Resource resource(ClientModel client) {
     @Override
     public Map<String, String> getPermissions(ClientModel client) {
         initialize(client);
+        //does it make sense to return the scopes if fine-grained permissions are disabled?
```
I don't think it does. The calling code does not call this method if the feature is disabled. See https://github.com/keycloak/keycloak/blob/2b996b12a180c426bbefe5cb44b6fce1af982840/services/src/main/java/org/keycloak/services/resources/admin/ClientResource.java#L660.

pedroigor commented on services/src/main/java/org/keycloak/services/resources/admin/permissions/GroupPermissions.java:
```diff
@@ -222,6 +222,7 @@ public Resource resource(GroupModel group) {
     @Override
     public Map<String, String> getPermissions(GroupModel group) {
         initialize(group);
+        //does it make sense to return the scopes if fine-grained permissions are disabled?
```
Same comment and code pattern from the previous comments on clients.

pedroigor commented on services/src/main/java/org/keycloak/services/resources/admin/permissions/IdentityProviderPermissions.java:
```diff
@@ -140,6 +141,7 @@ public Resource resource(IdentityProviderModel idp) {
     @Override
     public Map<String, String> getPermissions(IdentityProviderModel idp) {
         initialize(idp);
+        //does it make sense to fill the scopes when fine-grained permissions are disabled?
```
Same comment and code pattern from the previous comments on clients.

pedroigor commented on services/src/main/java/org/keycloak/services/resources/admin/permissions/MgmtPermissions.java:
```diff
@@ -203,6 +203,7 @@ public RealmModel adminsRealm() {
 
     @Override
     public RolePermissions roles() {
+        //Is there any reason why role permissions are always created new while the other permissions are kept?
```
I don't think so. We can do the same as others and see how it goes.

pedroigor commented on services/src/main/java/org/keycloak/services/resources/admin/permissions/RolePermissions.java:
```diff
@@ -91,6 +91,7 @@ private void disablePermissions(RoleModel role) {
     @Override
     public Map<String, String> getPermissions(RoleModel role) {
         initialize(role);
+        //does it make sense to initialize the scopes when fine-grained permissions are disabled?
```
Same comment and code pattern from the previous comments on clients.

pedroigor commented on services/src/main/java/org/keycloak/services/resources/admin/permissions/RolePermissions.java:
```diff
@@ -602,6 +604,7 @@ private String getMapCompositePermissionName(RoleModel role) {
         return MAP_ROLE_COMPOSITE_SCOPE + ".permission." + role.getId();
     }
 
+    //What is this? Did somebody roll over the keyboard?;) Apparently not used
```
Definitely :) Let's remove it.

pedroigor commented on services/src/main/java/org/keycloak/services/resources/admin/permissions/UserPermissions.java:
```diff
@@ -145,6 +145,7 @@ private void initialize() {
     @Override
     public Map<String, String> getPermissions() {
         initialize();
+        //does it make sense to initialize the scopes when fine-grained permissions are disabled?
```
Same comment and code pattern from the previous comments on clients.

pedroigor commented:
@sschu I'm really pleased to collaborate with you on these changes and your willingness to improve this area. Thanks a lot for your time and effort.

Despite the answers to some of your comments, the proposed changes LGTM and they should help us to create a baseline to perform more aggressive refactoring in the future.

sschu commented:
@pedroigor Thanks for the roses! ;) I am also happy to contribute. I implemented your suggestions. Should I then squash the commits if everything is fine or can it be automatically squash-merged?

pedroigor commented:
I can squash before merging.

sschu commented:
Hmm, the failing test ran fine locally.

sschu commented:
@pedroigor The test failure is most likely not related to my changes.
